pkg/controllers/allocation/scheduling/suite_test.go

```diff
@@ -368,6 +368,91 @@ var _ = Describe("NodeAffinity", func() {
 	})
 })
 
+var _ = Describe("Preferential Fallback", func() {
+	Context("Required", func() {
+		It("should not relax requirements", func() {
+			provisioner.Spec.Zones = []string{"test-zone-1"}
+			provisioner.Spec.InstanceTypes = []string{"default-instance-type"}
+			pod := test.UnschedulablePod()
+			pod.Spec.Affinity = &v1.Affinity{NodeAffinity: &v1.NodeAffinity{RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{NodeSelectorTerms: []v1.NodeSelectorTerm{
+				{MatchExpressions: []v1.NodeSelectorRequirement{
+					{Key: v1.LabelTopologyZone, Operator: v1.NodeSelectorOpIn, Values: []string{"invalid"}}, // Should not be relaxed
+				}},
+			}}}}
+			ExpectCreated(env.Client, provisioner)
+			ExpectCreatedWithStatus(env.Client, pod)
+			// Don't relax
+			ExpectReconcileSucceeded(ctx, controller, client.ObjectKeyFromObject(provisioner))
+			pod = ExpectPodExists(env.Client, pod.Name, pod.Namespace)
+			Expect(pod.Spec.NodeName).To(BeEmpty())
+			// Don't relax
+			ExpectReconcileSucceeded(ctx, controller, client.ObjectKeyFromObject(provisioner))
+			pod = ExpectPodExists(env.Client, pod.Name, pod.Namespace)
+			Expect(pod.Spec.NodeName).To(BeEmpty())
+		})
+		It("should relax terms", func() {
+			pod := test.UnschedulablePod()
+			pod.Spec.Affinity = &v1.Affinity{NodeAffinity: &v1.NodeAffinity{RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{NodeSelectorTerms: []v1.NodeSelectorTerm{
+				{MatchExpressions: []v1.NodeSelectorRequirement{
+					{Key: v1.LabelTopologyZone, Operator: v1.NodeSelectorOpIn, Values: []string{"invalid"}},
+				}},
+				{MatchExpressions: []v1.NodeSelectorRequirement{
+					{Key: v1.LabelTopologyZone, Operator: v1.NodeSelectorOpIn, Values: []string{"invalid"}},
+				}},
+				{MatchExpressions: []v1.NodeSelectorRequirement{
+					{Key: v1.LabelTopologyZone, Operator: v1.NodeSelectorOpIn, Values: []string{"test-zone-1"}},
+				}},
+			}}}}
+			ExpectCreated(env.Client, provisioner)
+			ExpectCreatedWithStatus(env.Client, pod)
+			// Remove first term
+			ExpectReconcileSucceeded(ctx, controller, client.ObjectKeyFromObject(provisioner))
+			pod = ExpectPodExists(env.Client, pod.Name, pod.Namespace)
+			Expect(pod.Spec.NodeName).To(BeEmpty())
+			// Remove second term
+			ExpectReconcileSucceeded(ctx, controller, client.ObjectKeyFromObject(provisioner))
+			pod = ExpectPodExists(env.Client, pod.Name, pod.Namespace)
+			Expect(pod.Spec.NodeName).To(BeEmpty())
+			// Success
+			ExpectReconcileSucceeded(ctx, controller, client.ObjectKeyFromObject(provisioner))
+			pod = ExpectPodExists(env.Client, pod.Name, pod.Namespace)
+			ExpectNodeExists(env.Client, pod.Spec.NodeName)
```

Comment:
Should you check that the node is in `test-zone-1`?

Reply:
I originally had these checks in, but it could actually be any of the default zones. I know that `invalid` is invalid, so it's enough to know that a node exists.